### PR TITLE
[Integration]: Serpstack api integration

### LINKED ISSUE #5641: [Integration]: Serpstack API Integration
### Short description of current behavior

The goal of this issue is to build a new handler that will utilize the Serpstack API to connect MindsDB to Serpstack. This integration will enable users to retrieve search engine results pages (SERPs) data from various search engines such as Google, Bing, Yahoo, and others, and use it for data analysis, search engine optimization (SEO) insights, and other automations within MindsDB.

## Resources :bulb:
:checkered_flag: To get started, please refer to the following resources:

* :books: [Building Application Handler docs](https://docs.mindsdb.com/contribute/app-handlers): This documentation will guide you through the process of building a custom handler for MindsDB. It contains essential information on the handler structure, methods, and best practices.

* :book: [Serpstack API docs](https://serpstack.com/documentation): This documentation contains all the endpoints and parameters that are available through the Serpstack API, including information on how to authenticate, make requests, and handle responses.

* :point_right: [Twitter Handler](https://github.com/mindsdb/mindsdb/tree/staging/mindsdb/integrations/handlers/twitter_handler): This existing handler connects MindsDB with the Twitter API, allowing users to automate tasks related to posting and analyzing tweets. You can use the Twitter Handler as a reference for building the Serpstack API integration.

* :point_right: [Binance API Handler](https://github.com/mindsdb/mindsdb/tree/staging/mindsdb/integrations/handlers/binance_handler): Another example of a successful MindsDB integration to the Binance API to make trade data available to use for model training and predictions.

## Next Steps :male_detective: :female_detective:
* Familiarize yourself with the Serpstack API and its functionalities.
* Create a new handler that connects MindsDB to Serpstack.
* Implement the core features mentioned above.
* Thoroughly test the integration to ensure proper functionality.
* Submit a pull request with the new Serpstack Integration handler.

## Hackathon Issue :loudspeaker:

MindsDB has organized a hackathon on Hashnode to bring ML to more applications.

Each hackathon issue can bring you rewards and prizes :trophy: .

Check out the [MindsDB Hackathon rules on Hashnode](https://hashnode.com/hackathons/mindsdb)!

## The https://github.com/mindsdb/mindsdb/labels/hashnode-hackhathon Label

We are happy to welcome you on board! Please take a look at the rules below if you are participating in the [HashnodeHackhathon](https://hashnode.com/hackathons/mindsdb).

 *   We are not assigning this type of issue to anyone. Multiple contributors can work on this, or even better, you can create a team and work together. Join our [Slack](https://mindsdb.com/joincommunity) to discuss more details. We will merge the best-quality PR.

 *   After you create your first PR in the MindsDB repository, please sign our CLA to become a MindsDB contributor. You can do that by leaving a comment in the Pull Request that contains the following: I have read the CLA Document and I hereby sign the CLA

Thank you for contributing to MindsDB :bear: !

### Video or screenshots

_No response_

### Expected behavior

_No response_

### How to reproduce the error

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by andersooi**
## Description

Created an integration for Serpstack API

Fixes #5641 

## Type of change

- [x] ⚡ New feature (non-breaking change which adds functionality)

## Verification Process

To ensure the changes are working as expected:

 - [x]   Test Location: Please refer to the loom recording attached below
 - [x]   Verification Steps: Implemented the integration and documented the process in a README.md file

## Additional Media:

- [x] I have attached a brief loom video or screenshots showcasing the new functionality or change.

[Loom Recording](https://www.loom.com/share/3bcaf48ee84e4359b5e95e7e5a9a7f85?sid=b8f0169d-0adc-47bd-a2f8-7d6aa86c885f)

## Checklist:

- [x] My code follows the style guidelines(PEP 8) of MindsDB.
- [x] I have appropriately commented on my code, especially in complex areas.
- [x] Necessary documentation updates are either made or tracked in issues.
- [ ] Relevant unit and integration tests are updated or added.

### CODE REVIEW

**Comment by ZoranPandovski** [comment]
recheckcla

**Comment by mindsdbadmin** [comment]
**CLA Assistant Lite bot:** <br/>Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://github.com/mindsdb/mindsdb/blob/main/assets/contributions-agreement/individual-contributor.md) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.<br/>
   - - -
   I have read the CLA Document and I hereby sign the CLA
   - - -
   <sub>You can retrigger this bot by commenting **recheck** in this Pull Request</sub>

**Comment by andersooi** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by andersooi** [comment]
Hi @ZoranPandovski, there seems to be the following test that is failing but am unsure of what to do on my end to resolve it. If possible, could you give me some guidance on how to go about resolving it?

image

File: mindsdb/integrations/handlers/serpstack_handler/serpstack_tables.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+        params = {'access_key': self.handler.access_key}
+        conditions = extract_comparison_conditions(query.where)
+        params.update({condition[1]: condition[2] for condition in conditions if condition[0] == '='})
+
+        if 'query' not in params:
+            raise ValueError('Query is missing in the SQL query')
+        if 'type' not in params and hasattr(self, 'default_type'):
+            params['type'] = self.default_type
+        api_response = requests.get(self.handler.base_url, params=params)
```
Comment (ZoranPandovski) [inline mindsdb/integrations/handlers/serpstack_handler/serpstack_tables.py:29]: Can we check for exceptions here?

File: mindsdb/integrations/handlers/serpstack_handler/serpstack_tables.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+        Args:
+            data (dict): The result data.
+
+        Returns:
+            dict: A dictionary containing the extracted data.
+        """
+        raise NotImplementedError("Subclasses must implement this method.")
+
+    def filter_columns(self, result: pd.DataFrame, query: ast.Select = None):
```
Comment (ZoranPandovski) [inline mindsdb/integrations/handlers/serpstack_handler/serpstack_tables.py:56]: Can we add a comments about what this does?

File: mindsdb/integrations/handlers/serpstack_handler/serpstack_handler.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            response.success = False
+
+        return response
+
+    def check_connection(self) -> StatusResponse:
+        """ Checks connection to Serpstack API"""
+        response = StatusResponse(False)
+
+        try:
+            self.connect()
```
Comment (ZoranPandovski) [inline mindsdb/integrations/handlers/serpstack_handler/serpstack_handler.py:106]: Maybe we don't need to call connect again but simply make a health-check request to serpstack?

**Review by ZoranPandovski — CHANGES_REQUESTED** [review]
Thanks @andersooi , don't worry about that failing action. I've added few minor comments

**Comment by andersooi** [comment]
Hi @ZoranPandovski I have made some changes based on the comments you gave. Do let me know if I need to make any additional edits or clarifications

**Comment by ZoranPandovski** [comment]
Great, just two flake warnings and this is good to merge https://github.com/mindsdb/mindsdb/actions/runs/10061406316/job/28044774189?pr=9381#step:6:75

**Review by MinuraPunchihewa — CHANGES_REQUESTED** [review]
Hey @andersooi,
Thank you very much for your contribution. This looks great!
There is just a couple of things I wanted to point out,
1. The connect() method of our API handlers is typically expected to return an instance of a client (package) that will allow us to interact with the API in question. Not an instance of StatusResponse (although this is incorrectly used as a type hint in some of our handlers). [Here](https://github.com/mindsdb/mindsdb/blob/bc102f4011837d289bca43d7978717f408818e6b/mindsdb/integrations/handlers/ms_teams_handler/ms_teams_handler.py#L55) is a good example of a correct implementation. Now, I understand that since you are using `requests` here, there is no client instance as such that can be returned. Therefore, you can choose to return something else; for instance the `base_url` and then you can use this within your `BaseResultsTable` like `base_url = self.handler.connect()`. It is completely up to you how you want to handle this, but in essence, what needs to be done is to remove the `StatusResponse` logic from `connect()`.
2. The `check_connection()` method is used to determine if a valid connection with the API can be established using the credentials that have been provided. I see that you are checking the health status of the API in this method, however, this does not check the validity of the API key provided. Instead what you can do is to call `connect()` within `check_connection()` and determine the validity of the connection from there. This method does indeed require a StatusResponse object to be returned and therefore, you can use error handling with the call to `connect()` in order to return the required response. [Here](https://github.com/mindsdb/mindsdb/blob/bc102f4011837d289bca43d7978717f408818e6b/mindsdb/integrations/handlers/ms_teams_handler/ms_teams_handler.py#L88) again, is an example of a more suitable implementation of this.

Finally, if you could add a couple of screenshots of this handler that would be amazing.

**Comment by andersooi** [comment]
Hi @MinuraPunchihewa I have made some changes based on your comments. Do let me know if there are any areas that require further edits.

I have attached a link to the loom recording of the integration usage in the pull request's description, [here](https://www.loom.com/share/3bcaf48ee84e4359b5e95e7e5a9a7f85?sid=b8f0169d-0adc-47bd-a2f8-7d6aa86c885f) it is for easier access.

Attached below is also a screenshot of the integration in use.

image

Thank you.

**Review by MinuraPunchihewa — APPROVED** [review]
This looks good to me now! Thanks, @andersooi!

**Review by ZoranPandovski — CHANGES_REQUESTED** [review]
@andersooi THank you, can you please just merge master so we avoid failing tests?

**Comment by andersooi** [comment]
@ZoranPandovski done!

**Pull request merged**